### ISSUE TITLE
Support eager port binding in ServerBuilder

### ISSUE THREAD

**Issue by mehrdada** [grpc/grpc#16403]
I am exploring an implementation of gRPC Python server on top of the C++ Server API. Most things work right out of the gate, but there is a difference between the C++ and Python APIs in when they actually bind to a port (and return the bound port number or fail): Python immediately binds on `grpc.Server.add_[in]secure_port` whereas C++ `ServerBuilder` only actually binds on `BuildAndStart()`, which makes it hard to construct the server. (In general, `AddListeningPort` API is a bit non-intuitive in C++ as the pointer you pass to it is not populated right after the call, but only after `BuildAndStart()` is invoked.)

Something like the following would help with the supported functionality, but may require moving things around in the way we construct `grpc::Server`:

```c++
int ServerBuilder::BindToEndpoint(const std::string& addr_uri, std::shared_ptr<ServerCredentials> creds)
```

(Interestingly, the desired behavior from Python perspective matches the `grpc::Server` internals well, and I considered inheriting a `PyServer` class from `grpc::Server` and instantiating it via its `protected` constructor, but realized it is clearly marked as "PRIVATE API and will become `private` pretty soon".)

[cl/210822191 is a temporary workaround]

**Comment by mehrdada** [comment]
friendly ping

**Comment by mehrdada** [comment]
less friendly ping

**Comment by vjpai** [comment]
@lidizheng @ericgribkoff @gnossen : does the Python team still need something like this?

**Comment by mehrdada** [comment]
Reopening as this is still a relevant issue with no good solution.

**Comment by jtattermusch** [comment]
We acknowledge this to be a valid feature request but since no one has come forward to implement this, we are going to close it for now while maintaining the label/Help wanted to indicate that this can be picked up as future work
